### DISCUSSION TITLE
Hardhat Full Course - Looking for feedback to make it even BETTER!

### DISCUSSION THREAD

**Discussion by PatrickAlphaC** [smartcontractkit/full-blockchain-solidity-course-py#802] [Announcements]
Hi all!

## Thank you!!!

🎊 It's been an AMAZING ride to see all this wonderful feedback, I'm so happy so many people are learning, and I'm BEYOND thrilled to see you all in the community!

## Looking for feedback on our next 16 hour video!

I've been working on getting the javascript edition of the hardhat starter kit up and going, and I have read a TON of all your feedback which I'm putting to work to make our next more amazing! (although, I ❤️ python, so I don't think any other edition will be more amazing, but I'll try my hardest!)

## Please respond in this thread

If you could please add feedback to this thread of what you'd like to see in the hardhat edition of this course, that would be amazing, some suggestions might be:

- "Patrick, you should split out the video into smaller videos"
- "I'd like to see more DAO lessons"
- "Front end is hard, can you focus on that more?"
- "Installing python was really difficult, could you make it easier?"

Or anything else you might have for suggestions! If you see a suggestion from someone else, please react to it with a 👍 or other emoji!

Thanks all for being here, and looking forward to your feedback!

I also just want again to give a huge thank you to @cromewar for stepping in and helping out so thoroughly, so if you bump into him, make sure to say thank you!

**Comment by ratewar** [comment]
Please add examples for NFT, DAO, Gaming in your new course or Metaverse. A basic understanding of these will help the developers pick up whats happening in web3.0 space.

**Reply by PatrickAlphaC** [reply]
Gaming!! yes of course!! DAOs will be in there for sure.

**Comment by 13zebras** [comment]
1. Definitely break up video and put into a playlist. It is very hard to come back to the video and remember exactly where you were. Release the videos as you complete them.
2. Focus on the details of hardhat. It can be a bit complex.
3. Explain a bit about Ethers.js.
4. Frontend: you create a basic scaffold. Let the user do the rest. If they don't know how, maybe a follow-up video later or point to a React course. 
5. NFT, DAO definitely. But you might look at some of the tools for building DAO's.
6. DEFI!! 
7. Build something with OpenZeppelin contracts. Get users comfortable with using audited contracts.
8. TESTING!!
9. IPFS would be good for section on NFTs
10. Have a few large projects to do that the viewer can go do on their own. 

Your python course is excellent. I just didn't want to learn python to do it, so didn't finish. Excited for hardhat course!! Will do it ASAP!

**Reply by PatrickAlphaC** [reply]
These are ALL going to be in there. Thank you!

**Reply by jacobbinnie** [reply]
I've just been adding a # note at the bottom of the current file I'm working on stating where I'm up to in the video. Personally, I'm happy with the learning structure and I actually enjoy the _slightly out of date_ content as it challenges you to figure out how to implement for different solidity versions etc.

Overall, I'm almost two months into the 16 Hour Solidity course and I'm on the final hour finishing up the react stuff. I don't think there's a more valuable resource for free on the internet for learning solidity and blockchain development as this. Amazing stuff.

Thanks for all your hard work @PatrickAlphaC. Legend.

**Reply by PatrickAlphaC** [reply]
I was so mad, the instant I release the video like half of the packages slightly updated 😔 Something that I'm also working on....

I'm so glad you're doing well!! You'll have to keep me posted on if you got a job in the space :)

**Comment by haidaralimasu** [comment]
All would be awesome but can you please make much more focus on DAO. It is just my personal wish

**Reply by PatrickAlphaC** [reply]
DAOs will be in there :)

**Comment by xremora** [comment]
Your last series was great. 

My only feedback/suggestion would be to include more of the diagrams (in the form of blocks) in slides or any other form of visualization (if possible) too. That being said, the code-first approach is 🔥. 

Thank you for all your efforts.

**Reply by PatrickAlphaC** [reply]
What do you mean? Diagrams to show what's going on under the hood?

**Reply by xremora** [reply]
Yes ser

**Comment by 0xBumstead** [comment]
Thanks a lot Patrick, your videos are great and close to perfection so it’s not easy to figure how to improve. I have only those two little details maybe:

Please wait a second when you’re done entering some code and before switching to another window (for example when switching between the py scripts and the contracts), so it allows the viewers to pause the video with the code on screen and catch up when we’re late.

For some of the most complicated solidity functions, it could be a good idea to write the test script first, pretty much like it was a test driven development. Maybe it’s just me but sometimes I was a bit lost when you were coding the solidity part and I had to wait for the test to fully understand what the function was supposed to do. You see el famoso “aaaaaaaah yes! Now I get it! Let’s go back 10 minutes in the video”.

Thanks again.
Cheers

**Reply by PatrickAlphaC** [reply]
> pause the video

Yes! I think I'm going to try to speak slower for this too. 

> Test Driven Development

Hmmmm. This is really good feedback. I'll think more on this....

**Comment by pan-efs** [comment]
Very good job @PatrickAlphaC and @cromewar !!!

- I would suggest to split the video in half. Backend and Frontend.
- More integration tests on testnets (i.e. kovan, rinkeby etc.).
- Spend some more time about IPFS and Pinata. 
- I think it's better to build an advanced smart contract using the best practices instead of building new and simple ones for each lesson. 

Best regards!

**Reply by PatrickAlphaC** [reply]
Thanks for the feedback!

**Comment by nicolasguasca1** [comment]
Hello Patrick. Thanks for all the content. I literally trust you every step of the way. I would love to have what you did for the brownie one but I would dig more into devops to provide a glimpse of what does it take to maintain a protocol in Real-life. Containers? Just wondering...

Hope you can read me lol. 

Best wishes!

**Reply by PatrickAlphaC** [reply]
ahhh this is something I probably won't get to :( 

This is a MUCH longer conversation ahah.

**Comment by FrostCorealis** [comment]
Love your videos!!  Wouldn't change too much, but here are a couple of ideas:

1) A quick tour through the Hardhat site for people who're new (and people who've poked around a little but didn't explore fully).

2) Specifics on setting up Visual Studio Code would be good for people who are just starting.  Once you're through it, the setup doesn't seem so bad, but when actually doing it, there's a lot of googling, trouble-shooting, and frustration.  It helps to know some common issues, how to fix them, and for beginners to know they're not alone in struggling with the setup.  Nuances of importing Open Zeppelin & Chainlink contracts (pathing, etc.) would be a helpful part of this.

3) Including how frontends call functions in smart contracts would  be great, as would some basic security precautions (don't upload your private key but also how to avoid writing functions that will make contracts vulnerable to exploits).

4) How to write run & deploy scripts, and what the various commands in the scripts do.

5) As for content, your DnD characters on the Chainlink site are great.  That's a fun project.  NFTs with VRF, gaming, and some DeFi would be wonderful.  Having some complexity to projects makes them more real + gets people thinking about possibilities.

Take care, and thank you so much for all you're doing!!  😎

**Reply by PatrickAlphaC** [reply]
1. Yes!
2. Yes!
3. Yes!
4. Yes!
5. Ahaha glad you liked it!
Thanks for the feedback!

**Comment by cromewar** [comment]
Thank you Sir, the python edition course is a game changing for everyone trying to start on Web3 and learn everything they need to succeed.
I read the comments above and they have good ideas, I'm sure the hardhat edition is going to be huge!

**Reply by shucks18** [reply]
Thank u @cromewar for all your help and contributions like seriously man!

**Comment by ace-brown** [comment]
Thanks for everything you did. I say these two down below plus working on NFTs more.
"Front end is hard, can you focus on that more?"
"Installing python was really difficult, could you make it easier?"

**Reply by PatrickAlphaC** [reply]
Yes!

**Comment by rholderfield** [comment]
How to use a hardhat for Avalanche C-chain deployment? Deep dive into Hardhat.config.js task and options configurations.

**Reply by PatrickAlphaC** [reply]
Working with Avax is the same as any EVM chain :) 

And yes to deep dive.

**Comment by lalittanna** [comment]
More of your jokes.

**Reply by PatrickAlphaC** [reply]
Don't tempt me

**Comment by clwest** [comment]
Maybe a section on wallets?  There has been a lot of talk about wallets lately!

**Reply by PatrickAlphaC** [reply]
Hmm... Wallets aren't a huugggee role in developing... I sort of don't care what wallet you use.

**Reply by clwest** [reply]
I meant building a multi sig wallet.

**Comment by ikhide** [comment]
Thank you @PatrickAlphaC for the previous course. I really loved the fact that a decent smart contract was made after most of the chapters and how awesome brownie bake is. Although I won't be going through this course as you have advised not to take multiple courses, I would watch the final chapter and I hope it would:
- Contain a production-grade dapp.
- Be written in typescript, not JS.
- Implement a DAO for governance.

**Reply by PatrickAlphaC** [reply]
1. Contain a production-grade dapp.
Maybe... I'd have to go through a full audit... We'll see

2. Typescript
We will try :) 

3. DAO
We will try :)

**Comment by rtraba** [comment]
Thank you for your great work!
1- use typechain
2- use upgredability hardhat plugin, or any proxy upgradable approach
3- use hardhat verify for publishing code on etherscan directly from the project
4- keep track of deployed contracts on the same repository
5- More Contracts Testing: 
5.1 Unit tests
5.2 mocking
5.3 integration tests
5.4 Smart Contract Fuzzing
5.6 EVM manipulation ... mooving blocks, changing time for mid complexity testing scenarios
5.7 bignumbers/decimals mid complexity assertions

**Comment by buixuandinh711** [comment]
English is not my native language so I have to use subtitle in the course. The auto-generated subtitle is pretty good, but with some proper nouns and commands it can't translate correctly. I think it would be better if you could edit the English subtitles so that more foreigners can access the course.
Thank you for the great course!

**Reply by PatrickAlphaC** [reply]
This is great feedback!

**Comment by tane-git** [comment]
Hey Patrick,

I am currently 6 hours into the 16-hour Solidity/Python Course after recently finishing a Coding Bootcamp focusing on JavaScript. Thank you for this course it is epic and I am excited to finish it!

**Positive Feedback:**
Epic course,
Enthusiastic, engaging and fast-paced (in a good way, viewers can always pause) 

**Growth Feedback:**

While adapting to Python wasn't too hard, I spent a lot of time resolving issues related to Python, setting up my environment for Python and just getting lots of random little bugs that you didn't get. I can see why people love Python but I am missing my JS.

1. In regards to how the JS course could be better, anything you could do to make the set-up part of the course easier would be great (for people who have little JS experience). 

As someone new to programming, one of the biggest hurdles to get over is getting your environment set up. It seemed like there where a few points in the course, when you had only just finished guiding us through a pretty major set-up and then you would instantly dive into the coding. It can be pretty disheartening when you are blazing into the coding and I am stuck here with a terminal full of errors. Just separating the **set-up** and **coding** parts of the course a little more, and reminding people where to go for help if they are getting errors. 

Splitting the course into smaller videos (as mentioned a lot already) will help this as the comment section of the video will be more relevant for helping you find solutions to weird little bugs that are coming up for you. Viewers can share the specific problems they are having better.

2. I would also be interested in you talking about security throughout the course whenever you introduce a new service/dependency etc. For example, when we use _Infura_ as our Node provider, what does this mean for our security and centralization. What are the alternatives (like running your own node?).  People new to programming may be surprised at how many services and packages we are pulling in to use, it can be pretty overwhelming! It would be interesting if whenever you introduce a new thing in the course, just to briefly mention if it affects our app's security/centralization at all and if so, how.

**Reply by PatrickAlphaC** [reply]
> While adapting to Python wasn't too hard, I spent a lot of time resolving issues related to Python, setting up my environment for Python and just getting lots of random little bugs that you didn't get. I can see why people love Python but I am missing my JS.

Thanks for this! This is something I'm currently battling with... JS is a _big_ monster. 

> Just separating the set-up and coding parts of the course a little more, and reminding people where to go for help if they are getting errors.

I love this. Yes!

> Security and tradeoffs

good call! 

Thank you for all this!!

**Comment by EcologicWind** [comment]
Hi @PatrickAlphaC  thank you for the great course. I've learned so far. My suggestion would be: since you have covered most of the basic concepts in this course, maybe shift the focus a little bit towards reacting with off-chain oracles, i.e. keepers since you're the expert and there is so few people talking about them, layer 2 networks and bridging contracts, DAOs, etc. 

I've learned so much so far and I look forward to the new course.

**Reply by EcologicWind** [reply]
so much*

**Reply by PatrickAlphaC** [reply]
Good thought!

**Comment by TRAVINIC** [comment]
True

Em ter, 1 de fev de 2022 23:53, Patrick Collins ***@***.***>
escreveu:

**Comment by n4n0b1t3** [comment]
It would be great to have a not-too-long refresher on how teams collaborate in the web3 environment. In my experience, it is the biggest hurdle for medium-skill developers to get involved or even find a job. Probably everyone agrees, that starting your own project is a good and fast way to learn new technology, even better to join a smaller project you can contribute to and be challenged by your peers. Building everything yourself reminds me of my web1 experience around 1998. The real challenge is 
- Where to start? How to choose a project which matches your expertise you can contribute to? Whom to ask?
- How to coordinate with others?
- Where to learn good practices using repositories?
- What to expect from the code peer-review process? 
- Understand the way teams decide on who will do what aspect of the work. 
- the structure and organizational structure of DAOs (already mentioned)

**Comment by shucks18** [comment]
Working with L2's !!

**Comment by EngineerFiaz** [comment]
Hardhat has compatibility issues with Windows  and performing  well on WSL

**Reply by EngineerFiaz** [reply]
![image](https://user-images.githubusercontent.com/37836666/217473001-a8ef409b-5d72-48a0-82a3-35593b483b30.png)
